consensus/src/ordered_broadcast/types.rs

```diff
@@ -0,0 +1,497 @@
+use super::Error;
+use bytes::{Buf, BufMut};
+use commonware_codec::{Encode, EncodeSize, Error as CodecError, FixedSize, Read, ReadExt, Write};
+use commonware_cryptography::{
+    bls12381::primitives::{
+        group::{Public, Signature},
+        ops,
+        poly::{self, PartialSignature},
+    },
+    Digest, Verifier,
+};
+use commonware_utils::{union, Array};
+
+/// Used as the [`Index`](crate::Supervisor::Index) type.
+/// Defines the current set of sequencers and validators.
+///
+/// This is not a single "View" in the sense of a consensus protocol, but rather a continuous
+/// sequence of views in-which the set of sequencers and validators is constant.
+pub type Epoch = u64;
+
+/// Used as the [`Automaton::Context`](crate::Automaton::Context) type.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct Context<P: Array> {
+    /// Sequencer's public key.
+    pub sequencer: P,
+
+    /// Sequencer-specific sequential height. Zero-indexed.
+    pub height: u64,
+}
+
+const CHUNK_SUFFIX: &[u8] = b"_CHUNK";
+const ACK_SUFFIX: &[u8] = b"_ACK";
+
+/// Returns a suffixed namespace for signing a chunk.
+pub fn chunk_namespace(namespace: &[u8]) -> Vec<u8> {
+    union(namespace, CHUNK_SUFFIX)
+}
+
+/// Returns a suffixed namespace for signing an ack.
+pub fn ack_namespace(namespace: &[u8]) -> Vec<u8> {
+    union(namespace, ACK_SUFFIX)
+}
+
+/// Chunk is a message generated by a sequencer that is broadcasted to all validators.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Chunk<P: Array, D: Digest> {
+    /// Sequencer's public key.
+    pub sequencer: P,
+
+    /// Sequencer-specific sequential height. Zero-indexed.
+    pub height: u64,
+
+    /// Digest of the payload.
+    pub payload: D,
+}
+
+impl<P: Array, D: Digest> Chunk<P, D> {
+    /// Create a new chunk with the given sequencer, height, and payload.
+    pub fn new(sequencer: P, height: u64, payload: D) -> Self {
+        Self {
+            sequencer,
+            height,
+            payload,
+        }
+    }
+}
+
+impl<P: Array, D: Digest> Write for Chunk<P, D> {
+    fn write(&self, writer: &mut impl BufMut) {
```

Comment:
I think i likely changed it since you started working on this, but generally we are using `buf` instead of `writer` or `reader`

Reply:
I've continued to use `reader`/`writer` (as that is what it is) but happy to migrate all of this PR (including `threshold-simplex` if that is the preference).

I thought of `Buf*` more as a trait that they would implement rather than a good name but 🤷